game/gameboard.py

```diff
@@ -116,6 +116,16 @@ def is_game_over(self):
     def retrieve_matrix(self):
         return self._matrix
 
+    def get_matrix_after_doing(self, function = lambda x: x):
```

Comment:
Sorry, but  I found this really confusing. You are using the same method to iterate / map the board and generating a copy of it. Plus the name when it has no argument is really weird (`board.get_matrix_after_doing()` - after doing what?). I would suggest:

- If you want to iterate the board, make it an iterable object (see e.g. http://stackoverflow.com/questions/19151/build-a-basic-python-iterator).
- If you want to map it to another thing (and it is not already an iterator) define a map function that forces the caller to send the lambda. 
- If you want the matrix, define a `get_matrix()`. But as we discussed before, I really think that defining a way to iterate on the object is a better approach.